### PR TITLE
Refactor context manager for sys path and current working directory in tests

### PR DESCRIPTION
**Pull request by Pierre-Sassoulas**
## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :hammer: Refactoring   |

## Description

There's still a lot to do there, but I'm going slowly. In particular I would have though that some pytest fixtures would have existed for these use cases, but it seem it's not the case, or is non-trivial.

### CODE REVIEW

File: setup.cfg [hunk 1]
```diff
@@ -13,6 +13,7 @@ test = pytest
 
 [tool:pytest]
 testpaths = tests
+pythonpath = data input regrtest_data
```
Comment (Pierre-Sassoulas) [inline setup.cfg:16]: This could be a standalone pull request, it's using pytest config to add fixture directory to the python path.

Comment (jacobtylerwalls) [reply setup.cfg:16]: I'm testing this change right now to see if that's all we need to revert the test adjustments in the astroid pr.

Comment (jacobtylerwalls) [reply setup.cfg:16]: Doesn't seem to make a difference -- which makes sense based on my limited understanding: the astroid pathways for defining modules are outside the control of pytest, and `tests` definitely is a package, we finally see it now as a namespace package.

Comment (jacobtylerwalls) [reply setup.cfg:16]: But please test and see if you get the same result! 🙏🏻

Comment (Pierre-Sassoulas) [reply setup.cfg:16]: (Slightly related we had a lot of issue with no``__init__.py`` in test directory, we could change that if it's causing more problem see https://github.com/PyCQA/pylint/issues/7085#issuecomment-1169863733)

Comment (jacobtylerwalls) [reply setup.cfg:16]: I think that's fixed in #7114, which has nothing to do with astroid thankfully, I just figured it out by reading the code comments :D

Comment (Pierre-Sassoulas) [reply setup.cfg:16]: I'm going to test on the 2.12 branch, I don't know why I did not do that earlier.

Comment (jacobtylerwalls) [reply setup.cfg]: If it helps, I noticed there were subtle behavior differences in running the tests depending on whether you have an editable install. (e.g. whether you're picking up `pylint/tests` or `astroid/tests` as `tests` automatically before you get to namespace package resolution)

Comment (Pierre-Sassoulas) [reply setup.cfg]: I removed the commit as I think you're right that astroid is not using the sys.path provided by pytest.

**Review by jacobtylerwalls — APPROVED** [review]

**Comment by DanielNoord** [comment]
This merge broke `pre-commit` on `main`. I think the runner image was updated on `pre-commit` and now probably runs Python 3.10.

@Pierre-Sassoulas We should probably also update the version with which we run `pre-commit` in CI for `pylint` and remove the useless ignores that are now being flagged.

**Comment by DanielNoord** [comment]
> Thanks for following that I missed it ! Well I happen to also have upgraded to python 3.10 last week so I think it's the right move. 

Officialy we support developing on `3.8` since that's supported by `pydocstringformatter` and necessary for the correct functional test output, but I think it makes sense to run our CI on `3.10` as it just frees up the runners sooner.

> The tests are SO MUCH FASTER locally, it's crazy (Maybe it was bloat that accumulated on my OS over the year or ubuntu 22.04).

🎉

**Comment by Pierre-Sassoulas** [comment]
Yes sure, we're not removing python 3.7 support, but LTS ubuntu is on 3.10 since april, and 3.10 is faster so.. let's run everything that run only once on 3.10 :)
